mesa_frames/abstract/space.py

```diff
@@ -733,3 +738,419 @@ def full_cells(self) -> DataFrame:
+        dimensions: Sequence[int]
+            The dimensions of the grid
+        torus : bool, optional
+            If the grid should be a torus, by default False
+        capacity : int | None, optional
+            The maximum number of agents that can be placed in a cell, by default None
+        neighborhood_type: str, optional
+            The type of neighborhood to consider, by default 'moore'.
+            If 'moore', the neighborhood is the 8 cells around the center cell (up, down, left, right, and diagonals).
+            If 'von_neumann', the neighborhood is the 4 cells around the center cell (up, down, left, right).
+            If 'hexagonal', the neighborhood are 6 cells around the center cell distributed in a hexagonal shape.
+        """
+        super().__init__(model, capacity)
+        self._dimensions = dimensions
+        self._torus = torus
+        self._cells_col_names = [f"dim_{k}" for k in range(len(dimensions))]
+        self._center_col_names = [x + "_center" for x in self._cells_col_names]
+        self._agents = self._df_constructor(
+            columns=["agent_id"] + self._cells_col_names, index_col="agent_id"
+        )
+        self._cells = self._df_constructor(
+            columns=self._cells_col_names + ["capacity"],
+            index_cols=self._cells_col_names,
+        )
+        self._offsets = self._compute_offsets(neighborhood_type)
+        self._grid_capacity = self._generate_empty_grid(dimensions, capacity)
+        self._neighborhood_type = neighborhood_type
+
+    def get_directions(
+        self,
+        pos0: GridCoordinate | GridCoordinates | None = None,
+        pos1: GridCoordinate | GridCoordinates | None = None,
+        agents0: IdsLike | AgentContainer | Collection[AgentContainer] | None = None,
+        agents1: IdsLike | AgentContainer | Collection[AgentContainer] | None = None,
+        normalize: bool = False,
+    ) -> DataFrame:
+        result = self._calculate_differences(pos0, pos1, agents0, agents1)
+        if normalize:
+            result = result / self._df_norm(result)
+        return result
+
+    def get_distances(
+        self,
+        pos0: GridCoordinate | GridCoordinates | None = None,
+        pos1: GridCoordinate | GridCoordinates | None = None,
+        agents0: IdsLike | AgentContainer | Collection[AgentContainer] | None = None,
+        agents1: IdsLike | AgentContainer | Collection[AgentContainer] | None = None,
+    ) -> DataFrame:
+        result = self._calculate_differences(pos0, pos1, agents0, agents1)
+        return self._df_norm(result)
+
+    def get_neighbors(
+        self,
+        radius: int | Sequence[int],
+        pos: GridCoordinate | GridCoordinates | None = None,
+        agents: IdsLike | AgentContainer | Collection[AgentContainer] | None = None,
+        include_center: bool = False,
+    ) -> DataFrame:
+        if __debug__:
+            if pos is not None and agents is not None:
```

Comment:
see #31. By default this is active, but if you run with -O flag, it doesn't even get compiled to bytecode. I think it might help improve performance overall, once development phase of the model is over and you know there are no bugs in your code.

Reply:
I see, though in this specific case, the `if type_is_such` checking actually constraints Mypy/Pyright when doing type checking, doesn't it. I think at least that the ones that are for typing, are not optional.

Reply:
I will remove the type checking altogether because I am already doing it in the concrete `get_neighborhood` method but you are right that it constrains static type checking. Sometimes this is not a problem (eg. see  the [`_get_df_coords`](https://github.com/adamamer20/mesa-frames/pull/32/files#diff-819a028fdb12dc7f03d0e51845dd9a3199a4d96708d3b0a8490cd7b2eea659c8R1054-R1110) method, if both `pos `and `agents ` are `None `then a ValueError is raised in the `else` block anyway), sometimes it might be.